tests/validate_yaml.py

```diff
@@ -7,27 +7,82 @@
 import yaml
 import sys
 
-def validate_yaml():
+def recursive_merge(d1, d2):
```

Comment:
We already have got a [similar procedure in `kernelci.config`](https://github.com/kernelci/kernelci-core/blob/2c249d6c9389e3c8f129afead12108421322b633/kernelci/config/__init__.py#L53).

Do you think it would be possible to refactor both these validators to use a common code base instead of separate check implementations?

Reply:
Yes, this is why i keep it as simple functions.
But tbh i dont want to touch kernelci.config overengineered stuff or waste time trying to integrate with it.
Maybe one day we will have budget to rework it.

Reply:
I'm afraid time/energy saving could only work short-term in this case - long-term there will be two implementations to maintain which should also be considered as an additional cost and an error-prone conditions (especially given the fact that this is data validation code).

What I meant was that extracted data merge code from the linked `kernelci.config` module to a separate procedure could also be reused here.

Reply:
At this particular case I'm confident i don't want to add dependency on kernelci-core. At least until we clean it up from legacy code and over-engineered parts.
If i add such dependency, it adds one big hell of dependencies and additional procedures instead of very basic verification.
That means i need to fetch kernelci-core repository, install it, with all dependencies it want, just for trivial yaml script?

If we go over proper way, such trivial functions should be made as separate library. And for example we can take a look to https://pypi.org/project/deepmerge/ which looks like does the job, instead of reinventing the wheel here and in kernelci-core. But again, we dont have time allocated for such research.

Additionally this merge function already expected be totally different from kernelci-core, as i'm planning to add more validation during merge as well, which kernelci-core likely wont have.
For example we might have job kbuild-arm64-test in pipeline.yaml, and someone by mistake might miss and add same job definition in job-chromeos.yaml. This will create very bad situation as both jobs will merge in not very predictable way, and to deal with it, i might customize further merge script to detect if there is same 2nd level definition keys have conflicts and etc.

Reply:
I understand your point about limiting introduced dependencies, especially for simple tasks.

I don't fully understand why all these checks are implemented separately from the tool that seems to have been designed especially for this use case: [`kci config validate`](https://github.com/kernelci/kernelci-core/blob/404c850547101bebca29fffb22cabf7de07387ba/kernelci/cli/config.py#L35)

Maybe at some point these validators [will get merged](https://github.com/kernelci/kernelci-pipeline/pull/643#issuecomment-2167794407) and [run in a container](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) including [all the necessary tools](https://hub.docker.com/r/kernelci/kernelci)

(marking this thread as resolved because the point where the discussion exceeded potential ROI has been reached)